Cram embed ref

jkbonfield opened the pull request:
Adds a cram2ref subcommand which extracts the embedded reference out of a CRAM file.  (Ie one that was built using `view -O cram,embed_ref=1`)

Fixes #723 (or at least partially).  A full fix would require whoever produced the assembly and aligned data against it to look for zero-coverage regions and produce synthetic reads to patch the holes so the consensus is 100% covered, but arguably if you align raw data against your assembly and find places with zero coverage then I'd argue the consensus there wasn't worth storing as it's apparently wrong!

jkbonfield commented:
NB also see https://github.com/jkbonfield/htslib/blob/cram_embed_ref2/cram/cram_encode.c#L2739-L2851

This is an htslib copy of two functions first implemented here.  I'll work on tidying up the API there some and then cull them here.

That code permits CRAM to embed references by dint of computing it on-the-fly (assuming MD:Z tag exists - todo: make sure it does!).

jmarshall commented on doc/samtools-reference.1:
```diff
@@ -0,0 +1,110 @@
+samtools reference
+.RB [ -q ]
+.RB [ -r
+.IR region "] "
```
```suggestion
.IR region ]
```

jmarshall commented on doc/samtools-reference.1:
```diff
@@ -0,0 +1,110 @@
+also produce an incomplete reference.  Unlike embedded reference, this
+reference may not be sufficient for decoding a CRAM file as the CRAM
+slice headers store the MD5sum of the portion of reference than spans
+that slice, but the slice may not have 100% coverage leadng to Ns in
```
```suggestion
that slice, but the slice may not have 100% coverage leading to Ns in
```

jmarshall commented on doc/samtools-reference.1:
```diff
@@ -0,0 +1,110 @@
+Write the FASTA records to \fIFILE\fR.  By default this is sent to stdout.
+
+.TP 8
+-BI "-@ " INT
+The number of BAM/CRAM decompression threads to use in addition tot he
```
```suggestion
.BI "-@ " INT
The number of BAM/CRAM decompression threads to use in addition to the
```

whitwham commented on reference.c:
```diff
@@ -0,0 +1,591 @@
+        // check b->core.tid and flush old seq.
+        if (b->core.tid != last_tid) {
+            if (last_tid >= 0)
+                dump_ref(h, iter, last_tid, ref, ref_len, outfp, 1);
```
Should dump_ref verbose always be 1 here?  The same question for the other dump_refs in this function.

jkbonfield commented:
Thanks.  I added a remove to fix the `M` option and honour `verbose`.  I also added more error checking to `dump_ref` calls.

I can squash and rebase this if you've finished review, but did a separate commit initially so you can review the fixes.

whitwham commented:
Look good to me.  If you squash it I will merge it.

jkbonfield commented:
Rebased and squashed.
